pty/pty_windows.go

```diff
@@ -135,19 +141,51 @@ func (p *ptyWindows) Close() error {
 	}
 	p.closed = true
 
-	ret, _, err := procClosePseudoConsole.Call(uintptr(p.console))
-	if ret < 0 {
-		return xerrors.Errorf("close pseudo console: %w", err)
+	// if we are running a command in the PTY, the corresponding *windowsProcess
+	// may have already closed the PseudoConsole when the command exited, so that
+	// output reads can get to EOF.  In that case, we don't need to close it
+	// again here.
+	if p.console != windows.InvalidHandle {
+		ret, _, err := procClosePseudoConsole.Call(uintptr(p.console))
+		if ret < 0 {
+			return xerrors.Errorf("close pseudo console: %w", err)
+		}
+		p.console = windows.InvalidHandle
 	}
 
-	_ = p.outputWrite.Close()
+	// We always have these files
 	_ = p.outputRead.Close()
 	_ = p.inputWrite.Close()
-	_ = p.inputRead.Close()
+	// These get closed & unset if we Start() a new process.
+	if p.outputWrite != nil {
+		_ = p.outputWrite.Close()
+	}
+	if p.inputRead.Close() != nil {
+		_ = p.inputRead.Close()
+	}
 	return nil
 }
 
 func (p *windowsProcess) waitInternal() {
+	defer func() {
+		// close the pseudoconsole handle when the process exits, if it hasn't already been closed.
+		// this is important because the PseudoConsole (conhost.exe) holds the write-end
+		// of the output pipe.  If it is not closed, reads on that pipe will block, even though
+		// the command has exited.
+		// c.f. https://devblogs.microsoft.com/commandline/windows-command-line-introducing-the-windows-pseudo-console-conpty/
+		p.pw.closeMutex.Lock()
+		defer p.pw.closeMutex.Unlock()
+		if p.pw.console != windows.InvalidHandle {
+			ret, _, err := procClosePseudoConsole.Call(uintptr(p.pw.console))
+			if ret < 0 && p.cmdErr == nil {
+				// if we already have an error from the command, prefer that error
+				// but if the command succeeded and closing the PseudoConsole fails
+				// then record that error so that we have a chance to see it
+				p.cmdErr = err
```

Comment:
It'd be nice to log the error whether it's kept or discarded

Reply:
It would be nice, but I don't have a logger, and I'd have to change all the init signatures to get one, so I decided on this.

Another alternative I considered is panicking---something pretty messed up at the OS level is happening if we get an error here.  We're pretty much guaranteed to crash the agent if we panic, so it's likely we'd get told about it.